src/node/kernel_notifications.h

```diff
@@ -25,6 +25,11 @@ class KernelNotifications : public kernel::Notifications
     void progress(const std::string& title, int progress_percent, bool resume_possible) override;
 
     void warning(const bilingual_str& warning) override;
+
+    void fatalError(const std::string& debug_message, const bilingual_str& user_message = {}) override;
+
+    //! Useful for tests, can be set to false to avoid shutdown on fatal error.
+    bool m_shutdown_on_fatal_error{true};
```

Comment:
> Maybe more ideally we could drop the bool by just having the test call BOOST_CHECK(ShutdownRequested()) and AbortShutdown() after it calls MaybeCompleteSnapshotValidation. This would make the test more realistic and add better coverage, but I didn't check if that worked.

This doesn't work, because calling `StartShutdown` without calling `InitShutdownState` beforehand triggers an `assert(0)`.

Reply:
> This doesn't work, because calling `StartShutdown` without calling `InitShutdownState` beforehand triggers an `assert(0)`.

I guess then question would be why can't the test framework call `InitShutdownState`? But this suggests that if we remove globals in shutdown code later (I don't know if that's required for the kernel), we should be able to get rid of the `m_shutdown_on_fatal_error` variable pretty easily later too.